### DISCUSSION TITLE
Define hole from DXF Improvement

### DISCUSSION THREAD

**Discussion by Xudong-Wang96** [Eomys/pyleecan#360] [Q&A]
Hello,
Thank you very much for developing and sustaining PYLEECAN, which helps a lot for my current thesis.

I'm now modeling PMSM torque ripples using PYLEECAN and need to import a specific rotor hole from a given .dxf file.
Some issues happen when using GUI to import DXF for machine design:

1. It gets an error message when importing, which is 
> 'Error while readng dxf file: an integer is required'
    the integer refers to the Line 177 of DXF_Hole.py, i.e.  
          `if self.selected_list[closest_id] == 0:  # Unselected to selected `
    the closet_id is not integer.

2. The .dxf file used is exported from FEMM sofware by a .fem file.
    When trying different  file, it's found that the line/arc in .dxf cause this error, which disapear when file only contains points.

3. Have you encounted this error before? what do you suggest for debug improvement?
    My idea is maybe some problems in the `self.line_list`, and I'm also curious how to debug this GUI function?

Thank you very much and looking forward to your reply.
BR
Xudong Wang

**Comment by BonneelP** [comment]
Hello,

Thank you for your interest in pyleecan. We are happy to know that you use it for your thesis. Feel free to open some issues/discussions if there are some topics that you want to share with the community ;)

Regarding your bug first some background questions: 
- Which version of pyleecan are you using ? 
- Which version of matplotlib and ezdxf are you using ? 
- Did the image of your machine appears in the Hole_DXF GUI ? 
- Can you share your dxf file with us or is it confidential ? 

Normally the code shouldn't create a "closest_id" which is not a integer so I'm a bit confused by your error :/
The principal of this GUI is to convert the dxf file to pyleecan Line/Arc objects to then display/select the pyleecan objects. But the conversion is not really strong (we managed to make it work for FEMM generated files but maybe some other sources can introduce errors) and we have some line (like spline) that are not supported. So the issue can be in how we read the file. If the machine is correctly drawn in the GUI then the problem is elsewhere since the dxf file is not used any more. If you can share your dxf file, I can try to import it on my side. 

To debug pyleecan code I personally use Windows IDE VS code (https://pyleecan.org/get.pyleecan.html) and a cloned repository of pyleecan on a python without pyleecan installed with pip.

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Hello,
I have step files of rotor and stator that I want to include in my design. Is it possible to import .stp files or should I convert to .dxf? Also, could you please tell me where I could find the information on how to import this ? I can't seem to find such an option on GUI  or on the website.

**Reply by BonneelP** [reply]
Hello,
Step file are not available for now sadly. But the code architecture should make this addition "easy": the dxf is read, converted to pyleecan lines and then all the GUI is not related to dxf so "only" the convertion step=>pyleecan lines would be required. 
For now you will have to convert your step file to dxf and you may also need to open / export your dxf with FEMM if you have some trouble importing your file pyleecan (we are thinking about automatizing the call to FEMM when pyleecan struggles with importing a dxf). 

Regarding where to find information about this import you can check our latest newsletter: https://mailchi.mp/aa37a343d865/pyleecan-news-import-from-dxf-star-of-slot-multi-simulation?e=36a7e6111b
(by the way you can subscribe here: http://eepurl.com/gYV_C1)

Best regards,
Pierre

**Reply by SebGue** [reply]
Hello Pierre, 
I have used GMSH (with Elmer 3D mag.) lately and I think we could also use it to do step import, i.e. one can access lines, arc, etc. really easy.
So it will also be compatible with linux.
BG S

**Reply by Lavanya-Bala** [reply]
> Hello,
> Step file are not available for now sadly. But the code architecture should make this addition "easy": the dxf is read, converted to pyleecan lines and then all the GUI is not related to dxf so "only" the convertion step=>pyleecan lines would be required.
> For now you will have to convert your step file to dxf and you may also need to open / export your dxf with FEMM if you have some trouble importing your file pyleecan (we are thinking about automatizing the call to FEMM when pyleecan struggles with importing a dxf).
> 
> Regarding where to find information about this import you can check our latest newsletter: https://mailchi.mp/aa37a343d865/pyleecan-news-import-from-dxf-star-of-slot-multi-simulation?e=36a7e6111b
> (by the way you can subscribe here: http://eepurl.com/gYV_C1)
> 
> Best regards,
> Pierre

Thank you for this information.

**Reply by Lavanya-Bala** [reply]
I converted my .stp into .dxf and tried to import into Pyleecan for specific magnet shape and the following error occurs. 

**Old machine version detected, Updating the Winding object
Unable to load dxftype=INSERT. Removing element from preview**

On the GUI pop, it says: Error while reading .dxf file, an integer is required. 

Could you please let me know what seems to be the issue here?

**Reply by BonneelP** [reply]
Hello,

Old machine version detected, Updating the Winding object
=> This message is not related to the dxf, it is when you load your machine (we have changed how we define the winding)

Unable to load dxftype=INSERT. Removing element from preview
=> When reading the dxf we convert line by line to pyleecan object. We don't handle all the line types so we can skip the import of some lines. What you can do is open your dxf file on FEMM and export it again to dxf. FEMM should convert all the lines to a format that pyleecan can handle.

For the last message I don't know, is it possible for you to share your dxf file ? Otherwise maybe you can download the code and check which line raises this error. Hopefully the FEMM convertion will also correct this error.

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Hello,

Thank you, using FEMM to export dxf again helped and for the 'integer required' issue, reverting to ezdxf 0.14.2 cleared it.
